### ISSUE TITLE
sql,sqlbase: maintain an index free-list on table descriptor

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#47981]
**Is your feature request related to a problem? Please describe.**

As we move towards constructing more indexes in schema changes and swapping them out after construction, we risk creating very high index ids. This is a bummer because index IDs are varint encoded and currently we almost always can fit that ID into a single byte. After we drop old indexes, we could easily reuse them are the relevant range has been cleared.

**Describe the solution you'd like**

Keep an index id freelist on the table descriptor and add to it after clearing the data in the GC Job.

**Describe alternatives you've considered**

Just letting the IDs get big. 

**Additional context**

Another place where a free list might be valuable is for descriptor IDs. We could use a table to store those free IDs.

**Comment by ajwerner** [comment]
Relates to #47989

**Comment by nvanbenschoten** [comment]
Another alternative here (which @ajwerner and I discussed last week) is to scan over the indexes in the `TableDescriptor` to find a free ID (once `NextIndexID` overflows) instead of maintaining a discrete freelist. The `Indexes` slice in `TableDescriptor` should already be sorted by ID, so finding an unused ID should be a constant space, linear time (w.r.t the number of active indexes) operation. This operation is run during a schema change, so it's not latency sensitive. Furthermore, these IndexDescriptors are 472 bytes large. If a table has so many indexes that iterating over them is cost-prohibitive, I think we'll have bigger problems when we pull the table's descriptor into memory.

Also, I think the freelist would need to do a similar search anyway, so does it even save us anything? I guess a large enough freelist could amortize the search cost.

**Comment by ajwerner** [comment]
> Also, I think the freelist would need to do a similar search anyway, so does it even save us anything? I guess a large enough freelist could amortize the search cost.

No reason such a free list should be serialized. We could just lazily initialize one on the ImmutableTableDescriptor. The way things are going we’re not going to be passing raw `*TableDescriptor`s around much. 

All good points, no reason to actually add such a free list to the table descriptor. I don’t think the same arguments apply to descriptor id generation.

**Comment by nvanbenschoten** [comment]
> I don’t think the same arguments apply to descriptor id generation.

Agreed. The use of a freelist in https://github.com/cockroachdb/cockroach/issues/48438 seems a lot more compelling.
